pytorch_lightning/loops/fit_loop.py

```diff
@@ -248,9 +248,17 @@ def teardown(self) -> None:
 
     def on_save_checkpoint(self) -> Dict:
         state_dict = super().on_save_checkpoint()
+
+        epoch_loop = getattr(self, "epoch_loop", None)
+        if not epoch_loop:
+            return state_dict
+
         # TODO: update has_completed to its proper value
-        if self.trainer.train_dataloader is not None:
-            state_dict["dataloader_state_dict"] = self.trainer.train_dataloader.state_dict(has_completed=False)
+        should_skip = epoch_loop.batch_progress.is_last_batch and epoch_loop.has_completed_training_batch
+        if self.trainer.train_dataloader is not None and not should_skip:
+            state_dict["dataloader_state_dict"] = self.trainer.train_dataloader.state_dict(
+                has_completed=epoch_loop.has_completed_training_batch
```

Comment:
shouldn't this pass `not should_skip`?

Reply:
replace by `should_save_dataloader_dict`